### PR TITLE
support unique index

### PR DESCRIPTION
**Pull request by zhoney**
implemented: #633

Change-Id: I753a81d0d47b0d7979563d5a9f41d368e06a5b33

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/IndexType.java [hunk 1]
```diff
@@ -77,14 +80,17 @@ public HugeType type() {
...
             default:
                 throw new AssertionError(String.format(
                           "Unknown index type '%s'", this));
         }
     }
 
     public boolean isString() {
-        return this == SECONDARY || this == SEARCH || this == SHARD;
+        return this == SECONDARY || this == SEARCH || this == SHARD ||
+               this == UNIQUE;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/IndexType.java]: balance the length of two lines

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendMutation.java [hunk 2]
```diff
@@ -97,6 +97,14 @@ private void optimizeUpdates(BackendEntry entry, Action action) {
...
                         throw incompatibleActionException(action, originAction);
                     } else {
+                        if (entry.type().isUniqueIndex() &&
+                            action == Action.APPEND &&
+                            originAction == Action.APPEND) {
+                            throw new IllegalArgumentException(String.format(
+                                      "Unique constraint conflict is found " +
+                                      "in tx between %s and %s",
+                                      entry, originItem.entry()));
+                        }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendMutation.java]: move to line 95

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 3]
```diff
@@ -248,6 +265,13 @@ private void updateIndex(IndexLabel indexLabel, Object propValue,
         }
     }
 
+    private boolean existUniqueValue(IndexLabel indexLabel, Object value) {
+        ConditionQuery query = new ConditionQuery(HugeType.UNIQUE_INDEX);
+        query.eq(HugeKeys.INDEX_LABEL_ID, indexLabel.id());
+        query.eq(HugeKeys.FIELD_VALUES, value);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: set limit(1)

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraTables.java [hunk 4]
```diff
@@ -615,6 +615,22 @@ public void insert(CassandraSessionPool.Session session,
         }
     }
 
+    public static class UniqueIndex extends SecondaryIndex {
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraTables.java]: add TODO: set field value as key and set element id as value

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 5]
```diff
@@ -158,22 +160,34 @@ protected void updateIndex(Id ilId, HugeElement element, boolean removed) {
...
             if (property == null) {
                 E.checkState(hasNullableProp(element, fieldId),
                              "Non-null property '%s' is null for '%s'",
                              this.graph().propertyKey(fieldId) , element);
-                // Not build index for record with nullable field
-                break;
+                allPropValues.add(INDEX_NULL_SYM);
+                hasNull = true;
+            } else {
+                E.checkArgument(property.value() != INDEX_NULL_SYM,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: equals

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java [hunk 6]
```diff
@@ -3289,6 +3289,63 @@ public void testAddVerticesWithUniqueIndexInTx() {
...
+        graph().addVertex(T.label, "user", "name", "", "city", "");
+        graph().tx().commit();
+        graph().addVertex(T.label, "user", "name", "", "age", 18);
+        graph().tx().commit();
+        graph().addVertex(T.label, "user", "city", "", "age", 18);
+        graph().tx().commit();
+        graph().addVertex(T.label, "user", "name", "");
+        graph().tx().commit();
+        Assert.assertThrows(IllegalArgumentException.class, () -> {
+            graph().addVertex(T.label, "user", "age", 18);
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexCoreTest.java]: conflict with line 3313

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 7]
```diff
@@ -158,22 +160,34 @@ protected void updateIndex(Id ilId, HugeElement element, boolean removed) {
...
+                allPropValues.add(INDEX_NULL_SYM);
+                hasNull = true;
+            } else {
+                E.checkArgument(!INDEX_NULL_SYM.equals(property.value()),
+                                "Illegal value of index property: '%s'",
+                                INDEX_NULL_SYM);
+                allPropValues.add(property.value());
+            }
+            if (!hasNull) {
+                firstNullField++;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java]: if(firstNullField == -1) firstNullField  = allPropValues.size()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 8]
```diff
@@ -402,15 +412,11 @@ private void checkRepeatIndex(SchemaLabel schemaLabel) {
             if (!this.hasSubIndex(old)) {
                 continue;
             }
-
-            /*
-             * If existed label is prefix of new created label,
-             * remove the existed label.
-             */
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: not adaptable for unique index

[the remark this answers is not in the document]
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: add more comments

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendMutation.java [hunk 9]
```diff
@@ -92,6 +92,13 @@ private void optimizeUpdates(BackendEntry entry, Action action) {
                     }
                     break;
                 case APPEND:
+                    if (entry.type().isUniqueIndex() &&
+                        originAction == Action.APPEND) {
+                        throw new IllegalArgumentException(String.format(
+                                  "Unique constraint conflict is found in tx " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendMutation.java]: in transaction

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 10]
```diff
@@ -474,22 +480,32 @@ private void checkRepeatShardIndex(SchemaLabel schemaLabel) {
...
             List<String> newFields = this.indexFields;
             List<String> oldFields = this.transaction.graph()
                                          .mapPkId2Name(old.indexFields());
-            // New created label can't be prefix of existed label
-            E.checkArgument(!CollectionUtil.prefixOf(newFields, oldFields),
-                            "Fields %s of new index label '%s' is prefix of " +
-                            "fields %s of existed index label '%s'",
-                            newFields, this.name, oldFields, old.name());
+            E.checkArgument(!predicate.test(newFields, oldFields),
+                            "Repeat new index label %s(%s) with fields %s " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java]: Repeated

**Pull request merged**